astropy/io/fits/_tiled_compression/tiled_compression.py

```diff
@@ -0,0 +1,566 @@
+            )
+            tile_data = _finalize_array(
+                tile_buffer,
+                hdu._header,
+                tile_shape=actual_tile_shape,
+                lossless=not quantize,
+            )
+
+            if "ZBLANK" in row.array.names:
+                zblank = row["ZBLANK"]
+            elif "ZBLANK" in hdu._header:
+                zblank = hdu._header["ZBLANK"]
+            else:
+                zblank = None
+
+            if zblank is not None:
+                blank_mask = tile_data == zblank
+
+            if quantize:
+                dither_method = DITHER_METHODS[hdu._header.get("ZQUANTIZ", "NO_DITHER")]
+                dither_seed = hdu._header.get("ZDITHER0", 0)
+                q = Quantize(
+                    row=(irow + dither_seed) if dither_method != -1 else 0,
+                    dither_method=dither_method,
+                    quantize_level=None,
+                    bitpix=hdu._header["ZBITPIX"],
+                )
+                tile_data = np.asarray(
+                    q.decode_quantized(tile_data, row["ZSCALE"], row["ZZERO"])
+                ).reshape(actual_tile_shape)
+
+            if zblank is not None:
+                if not tile_data.flags.writeable:
+                    tile_data = tile_data.copy()
+                tile_data[blank_mask] = np.nan
+
+        data[tile_slices] = tile_data
+        istart[-1] += tile_shape[-1]
+        for idx in range(data.ndim - 1, 0, -1):
+            if istart[idx] >= data_shape[idx]:
+                istart[idx] = 0
+                istart[idx - 1] += tile_shape[idx - 1]
+
+    return data
+
+
+def compress_hdu(hdu):
+    """
+    Compress the data in a `~astropy.io.fits.CompImageHDU`.
+
+    The input HDU is expected to have a uncompressed numpy array as it's
+    ``.data`` attribute.
+
+    Parameters
+    ----------
+    hdu : `astropy.io.fits.CompImageHDU`
+        Input HDU to compress the data for.
+
+    Returns
+    -------
+    nbytes : `int`
+        The number of bytes of the heap.
+    heap : `bytes`
+        The bytes of the FITS table heap.
+    """
+
+    if not isinstance(hdu.data, np.ndarray):
+        raise TypeError("CompImageHDU.data must be a numpy.ndarray")
+
+    _check_compressed_header(hdu._header)
+
+    # TODO: This implementation is memory inefficient as it generates all the
+    # compressed bytes before forming them into the heap, leading to 2x the
+    # potential memory usage.  Directly storing the compressed bytes into an
+    # expanding heap would fix this.
+
+    tile_shape = _tile_shape(hdu._header)
+    data_shape = _data_shape(hdu._header)
+
+    compressed_bytes = []
+    gzip_fallback = []
+    scales = []
+    zeros = []
+    zblank = None
+
+    irow = 0
+    istart = np.zeros(len(data_shape), dtype=int)
+
+    noisebit = _get_compression_setting(hdu._header, "noisebit", 0)
+
+    while True:
+
+        if hdu.data is None:
+            break
```

Comment:
How `hdu.data` can be None ?

Reply:
Removed as indeed it should never be needed.